src/networking.c

```diff
@@ -2322,7 +2322,7 @@ int processMultibulkBuffer(client *c) {
                 sdsIncrLen(c->querybuf,-2); /* remove CRLF */
                 /* Assume that if we saw a fat argument we'll see another one
                  * likely... */
-                c->querybuf = sdsnewlen(SDS_NOINIT,c->bulklen+2);
+                c->querybuf = sdsnewlen(SDS_NOINIT,c->bulklen+2 < PROTO_MBULK_BIG_ARG*2 ? c->bulklen+2:PROTO_MBULK_BIG_ARG*2);
```

Comment:
But it causes another side effect that we may need more realloc for c->querybuf.
IMHO, I would prefer to increase maxmemory, rather than degrade overall performance for it.

Reply:
Sure, there is no change in performance for the scenario where the command is less than 64kb.
IMO, when it is larger than 64kb, the IO overhead will be the performance bottleneck, and an extra malloc will not affect performance

Reply:
I‘m too naive :) this pr indeed caused noticeable performance regression.
benchmark command:
`./memtier_benchmark -s  IP -p PORT -t 5 -c 20  --hide-histogram  --key-maximum=1000 --test-time=30 --ratio=1:0 --d
ata-size=512000`
100KB down 23%
200KB down 15%
500KB down 5%
1000KB down 3%
Perhaps we can enlarge this limit value?

Reply:
> I would prefer to increase maxmemory, rather than degrade overall performance for it.

I think this has nothing to do with the maxmemory configuration, but with the remaining memory. If there is sufficient memory left, this problem will not be caused, but we will always waste/evict the `bulklen` of space, which could be 512MB.

Reply:
> Perhaps we can enlarge this limit value?

i don't think we can predict the most reasonable limit value.

> but we will always waste/evict the bulklen of space, which could be 512MB.
https://github.com/redis/redis/blob/c8226ae37875da285ebedeef78ced87666ca1be7/src/server.c#L732-L750

`c->querybuf` would be shrinked in the above, so it doesn't waste forever.

Reply:
@sundb hi, I realized that the above benchmark will cause redis to always predict success, without considering the additional performance loss caused by prediction failure. I redesigned the test case to make it closer to the actual use scenario, and the performance increased by 8%.
`./memtier_benchmark -s 10.86.115.60 -p 7001 -t 5 -c 20  --hide-histogram  --key-maximum=1000 --test-time=30 --ratio=1:0 --data-size-list=1024:80,102400:10,1024000:10`

Reply:
But redis can support >4GB single value for string, list, and other data types.
This means that the performance degradation is more pronounced when the data size is larger than the limit.
Perhaps we can wait for other guys to share opinions.

Reply:
> c->querybuf would be shrinked in the above, so it doesn't waste forever.
thank you for your patience, you are right, it do not waste forever, but before client shrinked the querybuf, the eviction has been triggered. I'd like to  introduce a test case to prove that this is a bug.
1. set mammemory 2GB, maxmemory policy is allkeys-lru
2. fill a large number of keys, so that used_memory reaches 1.3GB
```
127.0.0.1:7001> info memory keyspace
# Memory
used_memory:1479255992
used_memory_human:1.38G
used_memory_rss:1496068096
used_memory_rss_human:1.39G
used_memory_peak:1481380336
used_memory_peak_human:1.38G
used_memory_peak_perc:99.86%
used_memory_overhead:655085712
used_memory_startup:865960
used_memory_dataset:824170280
used_memory_dataset_perc:55.75%
allocator_allocated:1479682288
allocator_active:1480151040
allocator_resident:1504387072
total_system_memory:134942027776
total_system_memory_human:125.67G
used_memory_lua:31744
used_memory_vm_eval:31744
used_memory_lua_human:31.00K
used_memory_scripts_eval:0
number_of_cached_scripts:0
number_of_functions:0
number_of_libraries:0
used_memory_vm_functions:32768
used_memory_vm_total:64512
used_memory_vm_total_human:63.00K
used_memory_functions:184
used_memory_scripts:184
used_memory_scripts_human:184B
maxmemory:2147483648
maxmemory_human:2.00G
maxmemory_policy:allkeys-lru
allocator_frag_ratio:1.00
allocator_frag_bytes:468752
allocator_rss_ratio:1.02
allocator_rss_bytes:24236032
rss_overhead_ratio:0.99
rss_overhead_bytes:-8318976
mem_fragmentation_ratio:1.01
mem_fragmentation_bytes:16832856
mem_not_counted_for_evict:0
mem_replication_backlog:0
mem_total_replication_buffers:0
mem_clients_slaves:0
mem_clients_normal:1800
mem_cluster_links:0
mem_aof_buffer:0
mem_allocator:jemalloc-5.2.1
active_defrag_running:0
lazyfree_pending_objects:0
lazyfreed_objects:0

# Keyspace
db0:keys=13000001,expires=0,avg_ttl=0
```
3. write a string key with 500MB, then observe the number of keys
```
127.0.0.1:7001> info memory keyspace
# Memory
used_memory:2014553864
used_memory_human:1.88G
used_memory_rss:2007941120
used_memory_rss_human:1.87G
used_memory_peak:2552996232
used_memory_peak_human:2.38G
used_memory_peak_perc:78.91%
used_memory_overhead:654468112
used_memory_startup:865960
used_memory_dataset:1360085752
used_memory_dataset_perc:67.54%
allocator_allocated:2014687656
allocator_active:2016681984
allocator_resident:2040918016
total_system_memory:134942027776
total_system_memory_human:125.67G
used_memory_lua:31744
used_memory_vm_eval:31744
used_memory_lua_human:31.00K
used_memory_scripts_eval:0
number_of_cached_scripts:0
number_of_functions:0
number_of_libraries:0
used_memory_vm_functions:32768
used_memory_vm_total:64512
used_memory_vm_total_human:63.00K
used_memory_functions:184
used_memory_scripts:184
used_memory_scripts_human:184B
maxmemory:2147483648
maxmemory_human:2.00G
maxmemory_policy:allkeys-lru
allocator_frag_ratio:1.00
allocator_frag_bytes:1994328
allocator_rss_ratio:1.01
allocator_rss_bytes:24236032
rss_overhead_ratio:0.98
rss_overhead_bytes:-32976896
mem_fragmentation_ratio:1.00
mem_fragmentation_bytes:-6591992
mem_not_counted_for_evict:0
mem_replication_backlog:0
mem_total_replication_buffers:0
mem_clients_slaves:0
mem_clients_normal:1800
mem_cluster_links:0
mem_aof_buffer:0
mem_allocator:jemalloc-5.2.1
active_defrag_running:0
lazyfree_pending_objects:0
lazyfreed_objects:0

# Keyspace
db0:keys=12984561,expires=0,avg_ttl=0
``` 

We have 0.7GB of remaining memory, but observed about 15000 keys had been evicted

Reply:
> But redis can support >4GB single value for string, list, and other data types.
This means that the performance degradation is more pronounced when the data size is larger than the limit.

Not sure if I understand correctly, please correct me if i am wrong，just for clarifying the code. Even if we limit the anticipate maxsize for querybuf, we will  realloc the determined bulk length in these codes. So the overhead of this limit will be constant and performance recession will be smaller when the data value beacome bigger. 
```
            c->qb_pos = newline-c->querybuf+2;
            if (!(c->flags & CLIENT_MASTER) && ll >= PROTO_MBULK_BIG_ARG) {
                /* When the client is not a master client (because master
                 * client's querybuf can only be trimmed after data applied
                 * and sent to replicas).
                 *
                 * If we are going to read a large object from network
                 * try to make it likely that it will start at c->querybuf
                 * boundary so that we can optimize object creation
                 * avoiding a large copy of data.
                 *
                 * But only when the data we have not parsed is less than
                 * or equal to ll+2. If the data length is greater than
                 * ll+2, trimming querybuf is just a waste of time, because
                 * at this time the querybuf contains not only our bulk. */
                if (sdslen(c->querybuf)-c->qb_pos <= (size_t)ll+2) {
                    sdsrange(c->querybuf,c->qb_pos,-1);
                    c->qb_pos = 0;
                    /* Hint the sds library about the amount of bytes this string is
                     * going to contain. */
                    c->querybuf = sdsMakeRoomForNonGreedy(c->querybuf,ll+2-sdslen(c->querybuf));
                }
            }
            c->bulklen = ll;
        }

Reply:
@judeng This is exactly where I mentioned above the potential for performance degradation.
If we have allocated enough space by `c->querybuf = sdsnewlen(SDS_NOINIT,c->bulklen+2);`, we don't need to call sdsMakeRoomForNonGreedy again.

Reply:
I'm wondering about your test above, normally it should evict keys of 300M, but now it only evicts 100M, are you testing with this PR?

Reply:
I get it! And I agree with your point, when the length of next request is the same as the current request, and greater than 64KB, that is, when the prediction is successful , this pr will result in an additional `sdsMakeRoomForNonGreedy` call.
Conversely, if the length of the next request is much smaller than the current request, that is to say, the prediction fails, this pr can also save CPU and memory.
This looks like a trade-off, let's wait for others to comment.

Reply:
ping @soloestoy @oranagra, hope this didn't bother you, I'd like to get some help about how to better fix this problem, because this problem continues to happen in our production environment. thank you.

Reply:
i might be missing some details, but i think this edge case doesn't have to be addressed.

the `maxmemory` in redis isn't aiming to limit just the dataset size, but also some jitter buffers (like query buffers and argv arrays), these may grow and induce eviction or OOM, and the maxmemory config should be set in a way to allow their existence.
in this case if the maxmemory is set to 1gb, and the client operates on strings as large as half of that, it'll cause issues, but i think this is an edge case that we should not aim to handle any differently.

at the same time, redis also doesn't really aims to be strictly below the maxmemory, and there are cases (e.g. during command execution) that it'll allocate a temporary buffer, or copy a large argument from the argv array into the key, before releasing the argv array, in which case this will cause that (possibly large) argument to be allocated twice for a brief moment (and will not reflect in the peak memory metric).
i would argue that the peak memory metric aims to represent the steady state, and not the small spikes during command execution.

Reply:
thank you and agree with you, but there seems to be some gap with this bug. In this case, the client always predicts the length of the next command unlimitedly, resulting in additional memory and CPU overhead after the prediction fails. Fixing this not only reduces memory spikes but also helps performance (in actual use, we will not always predict success)

Reply:
but just removing this prediction is bad (for performance), and we don't have any other way for a better prediction (unless you can suggest one).
and either way, this is an edge case (happens when using arguments that are roughly the same scale as the memory limit)

Reply:
> but just removing this prediction is bad (for performance), and we don't have any other way for a better prediction (unless you can suggest one).

I adopted the simplest and crudest way which limit the maximum size of the forecast, and simulated the traffic in the real world for benchmark(1KB:100KB:1000KB=80:10:10), and the performance increased by about [8%](https://github.com/redis/redis/pull/11833#discussion_r1116708310). But for commands that are always 1MB in size, the performance drops by 3%.

Reply:
@filipecosta90 maybe you can provide some insight about it.
i.e. consistent argument sizes vs variable ones.

Reply:
hey,guys. I profiled the performance by using perf, and fixed the code to improve perfomance when reaching the limit, here is the new bench result:
bench command:`./memtier_benchmark -s IP-p PORT -t 5 -c 20 --hide-histogram  --key-maximum=1000 --test-time=60 --ratio=1:0 --data-size=XXXXX`
|version| 35KB | 66KB |100KB|1MB|1024:20,102400:80|1024:80,102400:20|
|:------:|:-----:|:-----:|:------:|:----:|:--------------------:|:-------------------:|
|unstable(61acf515)|49848|30206|21332|2468|25865|63187|
|this pr(61b60ddc)|49785|29578|21174|2670|26754|68070|

From the results, even at about 64KB, where the performance is expected to downgrade the most, there is actually no regression.